### PR TITLE
Watch project.assets.json files for nuget restore

### PR DESCRIPTION
**Pull request by dibarbet**
Previously, we were not reloading the project when a restore happened, until you actually modified a file.  This adds a file watcher for the `project.assets.json` file to tell us if a restore has happened and that the server needs to reload the project.

After:
```
[Trace - 11:15:07 AM] Sending notification 'workspace/didChangeWatchedFiles'.
Params: {
    "changes": [
        {
            "uri": "file:///c:/Users/dabarbet/source/repos/trash/MultipleTestProjects/XunitProject/obj/project.assets.json",
            "type": 1
        }
    ]
}
```

### CODE REVIEW

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LoadedProject.cs [hunk 1]
```diff
@@ -216,6 +219,27 @@ void UpdateProjectSystemProjectCollection<T>(IEnumerable<T> loadedCollection, IE
             if (newItems.Count != oldItemsCount)
                 logger.LogTrace(logMessage, projectFullPathWithTargetFramework, newItems.Count);
         }
+
+        void WatchProjectAssetsFile(ProjectFileInfo currentProjectInfo, IFileChangeContext fileChangeContext)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LoadedProject.cs:223]: I decided to only watch the actual msbuild property for ProjectAssetsFile here - I took a look at a bit of the O# and CPS code, and it looked they were only looking at that specific property (not for any project.assets.json file in obj/).

File: src/Workspaces/Core/MSBuild.BuildHost/MSBuild/Constants/PropertyNames.cs [hunk 2]
```diff
@@ -9,6 +9,7 @@ internal static class PropertyNames
         public const string AllowUnsafeBlocks = nameof(AllowUnsafeBlocks);
         public const string AppConfigForCompiler = nameof(AppConfigForCompiler);
         public const string BaseAddress = nameof(BaseAddress);
+        public const string BaseIntermediateOutputPath = nameof(BaseIntermediateOutputPath);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/MSBuild.BuildHost/MSBuild/Constants/PropertyNames.cs:12]: Did you add this but then never use it?

**Review by jasonmalinowski — APPROVED** [review]
Functionally fine, but you don't need the record since you can already grab the previous data to compare against.

File: src/Workspaces/Core/MSBuild.BuildHost/MSBuild/Constants/PropertyNames.cs (hunk 2 above)
Comment (dibarbet) [reply src/Workspaces/Core/MSBuild.BuildHost/MSBuild/Constants/PropertyNames.cs]: oops - was playing around with it and left this there

**Pull request merged**